larpandoracontent/LArThreeDReco/LArPfoMopUp/RecursivePfoMopUpAlgorithm.cc

```diff
@@ -0,0 +1,86 @@
+/**
+ *  @file   larpandoracontent/LArTwoDReco/LArClusterMopUp/RecursivePfoMopUpAlgorithm.h
+ *
+ *  @brief  Implementation file for the recursive pfo mop up algorithm that runs other algs.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArThreeDReco/LArPfoMopUp/RecursivePfoMopUpAlgorithm.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+StatusCode RecursivePfoMopUpAlgorithm::Run()
+{
+    std::vector<RecursivePfoMopUpAlgorithm::pfoMergeStats> mergeStatsVecBefore(GetPfoMergeStats());
+
+    for (unsigned int iter = 0; iter < m_maxIterations; ++iter)
+    {
+
+        for (auto const &mopUpAlg : m_mopUpAlgorithms)
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RunDaughterAlgorithm(*this, mopUpAlg));
+
+        std::vector<RecursivePfoMopUpAlgorithm::pfoMergeStats> mergeStatsVecAfter(GetPfoMergeStats());
+
+        if (std::equal(mergeStatsVecBefore.cbegin(), mergeStatsVecBefore.cend(), mergeStatsVecAfter.cbegin(), mergeStatsVecAfter.cend(),
+                RecursivePfoMopUpAlgorithm::pfoMergeStatsComp))
+            break;
+
+        mergeStatsVecBefore = std::move(mergeStatsVecAfter);
+    } // iter
+
+    return STATUS_CODE_SUCCESS;
+} // Run
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+std::vector<RecursivePfoMopUpAlgorithm::pfoMergeStats> RecursivePfoMopUpAlgorithm::GetPfoMergeStats() const
```

Comment:
Out of curiosity, why do you think this would be better to pass by reference in this case?

Reply:
Hi @etyley - Please feel free to leave it as it is. The suggestion to change to pass (then fill) by reference was because it's what we've done across a lot of the codebase, originally motivated by efficiency worries (constructing multiple large objects), but @absolution1 and I discussed this, checked a few details online, and it's clear that the compiler will be optimising this trivial case. Hope that helps! Thanks.

Reply:
Okay great, I had assumed return value optimisation would help us here so there should be no efficiency change by returning a reference. I will leave this as it is then.